File: src/main/java/lolapi/ApiRequestLimiter.java
package lolapi;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.function.Supplier;

public class ApiRequestLimiter {
    private final List<RequestsRateLimit> limits;

    protected ApiRequestLimiter(List<RequestsRateLimit> limits) {
        this.limits = Collections.unmodifiableList(limits);
    }

    public static ApiRequestLimiterBuilder build() {
        return new ApiRequestLimiterBuilder();
    }

    public void execute(Runnable runnable) {
        acquire();
        runnable.run();
    }

    public <T> T execute(Supplier<T> supplier) {
        acquire();
        return supplier.get();
    }

    public void acquire() {
        for (RequestsRateLimit limit : limits) {
            limit.acquire();
        }
    }

    public static class ApiRequestLimiterBuilder {
        private List<RequestsRateLimit> limits;

        protected ApiRequestLimiterBuilder() {
            limits = new ArrayList<>();
        }

        public ApiRequestLimiterBuilder add(int requestsCount, TimeUnit timeUnit, int unitsCount) {
            return add(requestsCount, timeUnit, unitsCount, true);
        }

        public ApiRequestLimiterBuilder add(int requestsCount, TimeUnit timeUnit, int unitsCount, boolean equable) {
            RequestsRateLimit limit = equable
                    ? new RequestsRateLimitEquable(requestsCount, timeUnit, unitsCount)
                    : new RequestsRateLimitNonEquable(requestsCount, timeUnit, unitsCount);
            limits.add(limit);
            return this;
        }

        public ApiRequestLimiter build() {
            Collections.sort(limits);
            return new ApiRequestLimiter(limits);
        }
    }
}

File: src/test/kotlin/ApiRequestLimiter.kt
import com.company.lolapi.ApiRequestLimiter
import org.junit.jupiter.api.Test
import org.opentest4j.AssertionFailedError
import java.util.concurrent.TimeUnit.SECONDS

class ApiRequestLimiter {

    @Test
    fun `Requests count in equable limit`() {
        val timeUnit = SECONDS
        val timeUnitCount = 1
        val requestsLimitCount = 3
        val limiter = ApiRequestLimiter.builder()
                .add(requestsLimitCount, timeUnit, timeUnitCount, true)
                .build()
        val limitTime = timeUnit.toMillis(timeUnitCount.toLong())

        val requestsTime = getRequestsTime(limiter, requestsLimitCount + 1)

        if (requestsTime <= limitTime)
            throw AssertionFailedError("", limitTime, requestsTime)
    }

    @Test
    fun `Requests count less than the quantity limit run faster than the non equable time limit`() {
        val timeUnit = SECONDS
        val timeUnitCount = 1
        val requestsLimitCount = 3
        val limiter = ApiRequestLimiter.builder()
                .add(requestsLimitCount, timeUnit, timeUnitCount, false)
                .build()
        val limitTime = timeUnit.toMillis(timeUnitCount.toLong())

        val requestsTime = getRequestsTime(limiter, requestsLimitCount - 1)

        if (requestsTime >= limitTime)
            throw AssertionFailedError("Requests time above or equal limit time:", limitTime, requestsTime)
    }

    @Test
    fun `Requests count more than the quantity limit run not faster than the non equable time limit`() {
        val timeUnit = SECONDS
        val timeUnitCount = 1
        val requestsLimitCount = 3
        val limiter = ApiRequestLimiter.builder()
                .add(requestsLimitCount, timeUnit, timeUnitCount, false)
                .build()
        val limitTime = timeUnit.toMillis(timeUnitCount.toLong())

        val requestsTime = getRequestsTime(limiter, requestsLimitCount + 1)

        if (requestsTime <= limitTime)
            throw AssertionFailedError("Requests time below or equal limit time:", limitTime, requestsTime)
    }

    private fun getRequestsTime(limiter: ApiRequestLimiter, requestsCount: Int): Long {
        val startTime = System.currentTimeMillis()
        for (i in 1..requestsCount) {
            limiter.acquire()
        }
        return System.currentTimeMillis() - startTime
    }
}